v0.5.0: Support for at-jobs, bug fixes.

jimbru opened the pull request:
- Upgrade to v0.5.0.
- Core interface changes for semantic consistency.
- Support for at-jobs.
- Fix tries = 0 for crons bug.
  
  Yay!

jimbru committed:
Version to 0.5.0-SNAPSHOT

...and update dependencies.

jimbru committed:
New core interface

Creates a clearer distinction between workers, jobs, registration and
scheduling. Consistent conventions for each.

jimbru committed:
Support for at-jobs

Implements "at-jobs": jobs that can be scheduled to run once at a
specific future time. Call `backtick.core/schedule-at` to use. Adds a
new database column to track jobs' run-at time.

jimbru committed:
Set tries count properly when enqueuing cron jobs

The backtick_queue.tries counter was not getting set to 1 when a job
from the backtick_cron table was (simultaneously) enqueued and started.
Fundamentally this is because there are 2 separate code paths for
starting jobs--this puts the cron path in sync with the ordinary
queue-pop path.

jimbru committed:
Version to 0.5.0

ctdean commented on resources/backtick/migrations/103-at-jobs.up.sql:
```diff
@@ -0,0 +1 @@
+ALTER TABLE backtick_queue ADD COLUMN run_at timestamp without time zone;
```
do we need an index too.  I assumed so, but you might know better.

jimbru replied:
Yeah no, I don't; an index is probably the right move.

ctdean commented:
No one like `cron` name.  Poor `cron`.

ctdean commented on src/backtick/engine.clj:
```diff
@@ -117,19 +119,21 @@
 
 ;; Not fault tolerant
 (defn- pop-cron-aux []
```
should rename this one also to `pop-recurring-aux`

jimbru replied:
Good point.

ctdean commented on src/backtick/core.clj:
```diff
@@ -15,35 +20,56 @@
+                                  worker)
+                 nil)
+      (string? v) v
+      :else worker)))
```
Can you explain this use case where we return the worker?  Seems like this might be the error case where the worked was the function name as a string.

jimbru replied:
It seems like `engine` wants a function reference in its `add` functions. Since the workers map is 2-way, this ensures that our `schedule*` functions can take _either_ the worker name or a function reference and abstract over that distinction. Not sure exactly what error case you're referring to.

ctdean replied:
gotcha, new feature!

Can we double check that all the args are right then?  So:

```
    ...
   (string? v) v
   (fn? worker) worker
   :else (do (log/error ...) nil)
  ...
```

jimbru replied:
Certainly

jimbru commented:
> No one like cron name. Poor cron.

I like the cron name, but I think we should reserve it for what you might call a recurring-at-job. In other frameworks I've seen "cron" refer specifically to jobs that are scheduled with a crontab-style string (e.g. `1 5 * * *`). What we were calling cron isn't actually that powerful and I think we should distinguish the two, since I expect that at some point we'd support both.

ctdean commented:
This all seems fine to me.  Make the changes you want and the ship it.

+1

jimbru committed:
Rename cron -> recurring

Make the code base internally consistent with the new interface naming
convention. What was formerly known as "cron" is now known as
"recurring".
